I think we still need the `finalResultType` in https://github.com/retronym/scala/commit/1a59c55ed5c6a0d53acd0259090cf57328748451#diff-e2cfc09e4d7c753232666c6e31bd1926R1525

Example:
```
case class A private (x: Int)
case class B private (x: Int)(y: Int)

class C {
  def f = A(1)
  def g = B(1)(2)
}
```

With 2.11.7:
```
➜  sandbox git:(pr5002) scalac11 -version
Scala compiler version 2.11.7 -- Copyright 2002-2013, LAMP/EPFL
➜  sandbox git:(pr5002) scalac11 Test.scala
Test.scala:5: error: constructor A in class A cannot be accessed in class C
  def f = A(1)
          ^
Test.scala:6: error: constructor B in class B cannot be accessed in class C
  def g = B(1)(2)
          ^
two errors found
``


➜  sandbox git:(pr5002) ✗ git diff | cat
➜  sandbox git:(pr5002) qsc -version
Scala compiler version 2.11.8-20160302-221155-fe5bd09861 -- Copyright 2002-2016, LAMP/EPFL
➜  sandbox git:(pr5002) qsc Test.scala
Test.scala:6: error: constructor B in class B cannot be accessed in class C
  def g = B(1)(2)
          ^
one error found
```

--- a/src/compiler/scala/tools/nsc/typechecker/RefChecks.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/RefChecks.scala
@@ -1522,7 +1522,7 @@ abstract class RefChecks extends InfoTransform with scala.reflect.internal.trans
         sym.isCase &&
         sym.name == nme.apply &&
         isClassTypeAccessible(tree) &&
-        !tree.tpe.resultType.typeSymbol.primaryConstructor.isLessAccessibleThan(tree.symbol)
+        !tree.tpe.finalResultType.typeSymbol.primaryConstructor.isLessAccessibleThan(tree.symbol)
     }
 
     private def transformCaseApply(tree: Tree) = {
@@ -1571,8 +1571,15 @@ abstract class RefChecks extends InfoTransform with scala.reflect.internal.trans
         // term should have been eliminated by super accessors
         assert(!(qual.symbol.isTrait && sym.isTerm && mix == tpnme.EMPTY), (qual.symbol, sym, mix))
 
-      // SI-9546 isHigherKinded excludes generic case classes which are instead considered when transforming
-      //         the enclosing `TypeApply`.
+      // Rewrite eligible calls to monomorphic case companion apply methods to the equivalent constructor call.
+      //
+      // Note: for generic case classes the rewrite needs to be handled at the enclosing `TypeApply` to transform
+      // `TypeApply(Select(C, apply), targs)` to `Select(New(C[targs]), <init>)`. In case such a `TypeApply`
+      // was deemed ineligible for transformation (e.g. the case constructor was private), the refchecks transform
+      // will recurse to this point with `Select(C, apply)`, which will have a type `[T](...)C[T]`.
+      //
+      // We don't need to perform the check on the Select node, and `!isHigherKinded will guard against this
+      // redundant (and previously buggy, SI-9546) consideration.
       if (!tree.tpe.isHigherKinded && isSimpleCaseApply(tree)) {
         transformCaseApply(tree)
       } else {

--- a/test/files/run/t9546e.scala
+++ b/test/files/run/t9546e.scala
@@ -1,0 +1,15 @@
+case class A private (x: Int)
+case class B private (x: Int)(y: Int)
+
+class C {
+  def f = A(1)
+  def g = B(1)(2) // was: constructor B in class B cannot be accessed in class C
+}
+
+object Test {
+  def main(args: Array[String]): Unit = {
+    new C().f
+    new C().g
+  }
+
+}

